history-network.md

```diff
@@ -0,0 +1,300 @@
+
+All `content-key` values are encoded and decoded according to SSZ sedes.
+
+#### Block Header
+
+```
+content-key  = Container(chain-id: uint16, content-type: uint8, block-hash: Bytes32)
+content-type = 0x01
+```
+
+#### Block Body
+
+```
+content-key  = Container(chain-id: uint16, content-type: uint8, block-hash: Bytes32)
+content-type = 0x02
+```
+
+#### Receipts
+
+```
+content-key  = Container(chain-id: uint16, content-type: uint8, block-hash: Bytes32)
+content-type = 0x03
+```
+
+### Radius
+
+We define a `distance` function that maps a `node-id` and `content-id` pair to a 256-bit unsigned integer identically to the `distance` function for pairs of `node-id` values.
+
+Each node specifies a `radius` value, a 256-bit unsigned integer that represents the data that a node is "interested" in.
+
+```
+interested(node, content) = distance(node.id, content.id) <= node.radius
+```
+
+A node is expected to maintain `radius` information for each node in its local node table. A node's `radius` value may fluctuate as the contents of its local key-value store change.
+
+### Protocol Messages
+
+All protocol messages are transmitted via the `TALKREQ` and `TALKRESP` messages defined by the Discovery v5 protocol.
+
+Below is the general message encoding for all `TALKREQ` messages.
+
+```
+protocol              = portal-history
+request               = protocol-message-type || protocol-message
+protocol-message-type = uint8
+protocol-message      = byte array
+```
+
+Below is the general message encoding for all `TALKRESP` messages.
+
+```
+response              = protocol-message-type || protocol-message
+protocol-message-type = uint8
+protocol-message      = byte array
+```
+
+All `protocol-message` values are encoded and decoded according to SSZ sedes.
+
+If the size of a message would exceed the max packet size of 1280 bytes, then the transmission of the message contents would occur over [uTP](https://github.com/ethereum/stateless-ethereum-specs/blob/master/discv5-utp.md).
+
+uTP connections are initiated with randomly generated connection IDs. Upon sending a message with some `connection-id`, the sender should initiate a uTP stream using `connection-id`. Upon receiving a message that contains some `connection-id`, the recipient should listen for an incoming uTP connection with `connection-id`.
+
+NOTE: The `Offer`/`Accept`/`Store` messages do not conform to the request/response paradigm of Discovery v5. We plan to propose a change to the base protocol definition to loosen the language.
+
+#### Ping (0x01)
+
+Communicate `radius` information of the sender, and request `radius` information of the recipient.
+
+```
+protocol-message-type = 0x01
+protocol-message      = Container(radius: uint256)
+```
+
+#### Pong (0x02)
+
+In response to a `Ping` message, communicate `radius` information of the sender.
+
+```
+protocol-message-type = 0x02
+protocol-message      = Container(radius: uint256)
```

Comment:
I'm wondering if we should also include ENR `sequence_number` here as well so that we don't have to somehow fall back to base protocol messages in order to ensure we have an up-to-date picture of a nodes ENR.